commit/solvers.py

```diff
@@ -373,10 +375,171 @@ def fista( y, A, At, tol_fun, tol_x, max_iter, verbose, x0, omega, proximal) :
+    else:    
+        L = ( np.linalg.norm( A.dot(grad) ) / np.linalg.norm(grad) )**2
+    mu = 1.9 / L
+
+    # Main loop
+    if verbose >= 1 :
+        print()
+        print( "      |  1/2||Ax-y||^2      Omega      |  Cost function    Abs error      Rel error    |      Abs x          Rel x    " )
+        print( "------|--------------------------------|-----------------------------------------------|------------------------------" )
+    iter = 1
+    while True :
+        if verbose >= 1 :
+            print( "%4d  |" % iter, end="" )
+            sys.stdout.flush()
+
+        # Smooth step
+        x = xhat - mu*grad
+
+        # Non-smooth step
+        proximal( x )
+        reg_term_x = omega( x )
+
+        # Check stepsize
+        tmp = x-xhat
+        q = qfval + np.real( np.dot(tmp,grad) ) + 0.5/mu * np.linalg.norm(tmp)**2 + reg_term_x
+        if sqrt_W is not None:
+            res = sqrt_W * ( A.dot(x) - y )
+        else:
+            res = A.dot(x) - y
+        res_norm = np.linalg.norm(res)
+        curr_obj = 0.5 * res_norm**2 + reg_term_x
+
+        # Backtracking
+        while curr_obj > q :
+            # Smooth step
+            mu = beta*mu
+            x = xhat - mu*grad
+
+            # Non-smooth step
+            proximal( x )
+            reg_term_x = omega( x )
+
+            # Check stepsize
+            tmp = x-xhat
+            q = qfval + np.real( np.dot(tmp,grad) ) + 0.5/mu * np.linalg.norm(tmp)**2 + reg_term_x
+            if sqrt_W is not None:
+                res = sqrt_W * ( A.dot(x) - y )
+            else:
+                res = A.dot(x) - y
+            res_norm = np.linalg.norm(res)
+            curr_obj = 0.5 * res_norm**2 + reg_term_x
+
+        # Global stopping criterion
+        abs_obj = abs(curr_obj - prev_obj)
+        rel_obj = abs_obj / curr_obj
+        abs_x   = np.linalg.norm(x - prev_x)
+        rel_x   = abs_x / ( np.linalg.norm(x) + eps )
+        if verbose >= 1 :
+            print( "  %13.7e  %13.7e  |  %13.7e  %13.7e  %13.7e  |  %13.7e  %13.7e" % ( 0.5 * res_norm**2, reg_term_x, curr_obj, abs_obj, rel_obj, abs_x, rel_x ) )
+
+        if abs_obj < eps :
+            criterion = "Absolute tolerance on the objective"
+            break
+        elif rel_obj < tol_fun :
+            criterion = "Relative tolerance on the objective"
+            break
+        elif abs_x < eps :
+            criterion = "Absolute tolerance on the unknown"
+            break
+        elif rel_x < tol_x :
+            criterion = "Relative tolerance on the unknown"
+            break
+        elif iter >= max_iter :
+            criterion = "Maximum number of iterations"
+            break
+
+        # FISTA update
+        t = 0.5 * ( 1 + sqrt(1+4*told**2) )
+        xhat = x + (told-1)/t * (x - prev_x)
+
+        # Gradient computation
+        if sqrt_W is not None:
+            res = sqrt_W * ( A.dot(xhat) - y )
+        else:
+            res = A.dot(xhat) - y
+        xarr = np.asarray(x)
```

Comment:
As you're doing this PR, will you remove this unused variable?

Reply:
Yes, I did